### ISSUE TITLE
Behavioural vs. Object Oriented

### ISSUE THREAD

**Issue by AmitKumarDas** [AmitKumarDas/Decisions#25]
#### Is your code behavioural than OO:
- If .go file uses a lot of functions than methods ?
- If .go file uses/creates objects that always end with -er ?
  - Using -er named objects is the easy way. But lets the code open for modification.
- Do you start thinking about Interfaces than about Objects?
  - This is a classic text book suggestion, but IMHO this is WRONG.

#### Alternatively checks if your code is NOT DRY?
- If the code is procedural, it leads to lot of code
  - This can be minimised by modularizing. However, it just moves the problem.
  - Modularization cannot be equated with being DRY.
- Are their functions or methods that access properties from Obj A & mutate properties
into Obj B?
  - Access means get calls
  - Mutate means set calls
- If features use YAML & JSON as intent but they don't use marshall & un-marshall code
- If code does not use templates to display the objects

### Summary
_What if I don't understand any of the above? Can you give me one single formula to write code that lasts?_

#### Here it is:
- Write unit tests
- Never use third party mocks
  - This provides quick workaround to hide the issues of your behaviorial design
- Do not go about writing interfaces to implement UT
  - In other words create an Interface & then extend the interface with:
    - 1. Original code &
    - 2. Mock code
  - NOTE: This is too much of code & defeats the DRY principle
- Check if the language provides method overriding features to mock
  - How about monkey patching to mock? 
  - I will be using it this time, since I tried all other tricks & they are not sustainable

### Now some quick tests?
- Test above learnings by reviewing below PR:
ref - https://github.com/openebs/maya/pull/245/files
